src/edgepi/peripherals/gpio.py

```diff
@@ -45,6 +45,67 @@ def read_state(self):
         """
         return self.gpio.read()
 
+    def open_read_state(self, pin_num:int, pin_dir:str, pin_bias:str):
+        """
+        To minimize issues with the lock, we open & read in a single function call
+        """
+        try:
+            # pylint: disable=consider-using-with
+            GpioDevice.lock_gpio.acquire()
+            gpio   = GPIO(self.gpio_fd, pin_num, pin_dir, bias=pin_bias)
+            result = gpio.read()
+
+        finally:
+            try:
+                gpio.close()
+            except Exception as exc:
+                raise OSError(f"Failed to close {self.gpio_fd}") from exc
+            finally:
+                GpioDevice.lock_gpio.release()
+
+        return result
+
+    def open_read_state_batch(
+        self,
+        pin_num_list: list[int],
+        pin_dir: str,
+        pin_bias: str,
+    ) -> list:
+        """
+        Batch several gpio reads into a single lock / unlock. We can also take advantage of
+        the fact we're accessing the gpio only from different pins, and so we can open & close
+        just a single time.
+        """
+        results = []
+        try:
+            # pylint: disable=consider-using-with
+            GpioDevice.lock_gpio.acquire()
+            gpio = None
+            try:
+                for pin_num in pin_num_list:
+                    if gpio is None:
+                        gpio = GPIO(self.gpio_fd, pin_num, pin_dir, bias=pin_bias)
+                    else:
+                        # NOTE: we'll need to be careful when we update periphery, since we depend
+                        # on private functionality
+                        gpio._line = pin_num # pylint: disable=protected-access
+                        # pylint: disable=protected-access
+                        gpio._reopen(
+                            pin_dir, edge="none", bias=pin_bias, drive="default", inverted=False
+                        )
+                    results += [gpio.read()]
```

Comment:
Is the result of `gpio.read()` larger than 1?

Reply:
[`gpio.read()`](https://python-periphery.readthedocs.io/en/latest/gpio.html#periphery.GPIO.read) returns a scalar boolean

I've updated the types hints for functions in this module https://github.com/EdgePi-Cloud/edgepi-python-sdk/pull/423/commits/654ff9f98798cedfbdd875bc2db024c347e73b36

Reply:
How about `results.append(gpio.read()`? It avoids creating the new list to temporarily store `gpio.read()`.

Reply:
`+= []` is common pythonic way for appending to a list, and I usually prefer it for readability since it's shorter. 

However looking into it, it seems `append` is a little bit more performant. Or rather, it's more that `+=` can be harder for python to optimize (and sometimes [becomes quadratic](https://google.github.io/styleguide/pyguide.html#:~:text=accumulating%20a%20string%20with%20addition%20can%20lead%20to%20quadratic%20rather%20than%20linear%20running%20time) in the worst case!!). I doubt it will make a performance different in this case, but it's interesting to know that the google style guide prefers `append`!

https://github.com/EdgePi-Cloud/edgepi-python-sdk/pull/423/commits/f4a4a4f0e6500de82e335263d40125ec8eca320d